1235 angr on android (#1351)
OWASP/owasp-mstg: Document/0x05b-Basic-Security_Testing.md

##### Angr

> Angr is a Python framework for analyzing binaries. It is useful for both static and dynamic symbolic ("concolic") analysis. In other words: given a binary and a requested state, Angr will try to get to that to the requested state in the binary, using formal methods to find a path, as well as brute forcing. Using angr to get to the requested state is way faster than taking manual steps for debugging and searching the path towards the required state. Angr operates on the VEX intermediate language and comes with a loader for ELF/ARM binaries, so it is perfect for dealing with native code, such as native Android binaries.

Reviewer: ```suggestion
Angr is a Python framework for analyzing binaries. It is useful for both static and dynamic symbolic ("concolic") analysis. In other words: given a binary and a requested state, angr will try to get to the requested state in the binary, using formal methods (a technique used for static code analysis) to find a path, as well as brute forcing. Using angr to get to the requested state is way faster than taking manual steps for debugging and searching the path towards the required state. Angr operates on the VEX intermediate language and comes with a loader for ELF/ARM binaries, so it is perfect for dealing with native code, such as native Android binaries.
```

##### Angr

> Angr is a Python framework for analyzing binaries. It is useful for both static and dynamic symbolic ("concolic") analysis. In other words: given a binary and a requested state, Angr will try to get to that to the requested state in the binary, using formal methods to find a path, as well as brute forcing. This can make reverse-engineerig "I just want a solution style" - way easier than with for instance IDA Pro, Radare2, or Ghidra. Angr operates on the VEX intermediate language and comes with a loader for ELF/ARM binaries, so it is perfect for dealing with _native_ Android binaries.

Reviewer: - This can make reverse-engineerig "I just want a solution style" - way easier than with for instance IDA Pro, Radare2, or Ghidra. -> two things: 1) I would rephrase the first sentence but don't know why now. 2) why is it easier?

- Angr operates on the VEX intermediate language and comes with a loader for ELF/ARM binaries, so it is perfect for dealing with _native_ Android binaries. -> do you mean Android native code? Maybe: "Android native code such as native libraries"?

Author: 1.: Is `This can make reverse-engineerig "I just want a solution style" - way easier than with for instance IDA Pro, Radare2, or Ghidra.`  --> `Using angr to get to the requested state is way faster than taking manual steps for debugging and searching the path towards the required state.` better?
2.: yes, i meant Android native code such as native libraries. Will fix that

Reviewer: Yes, I like that!

> Angr allows for disassempbly, program instrumentation, symbolic execution, control-flow analysis, data-dependency analysis, decompilation and more, given a large set of plugins.

Reviewer: ```suggestion
Angr allows for disassembly, program instrumentation, symbolic execution, control-flow analysis, data-dependency analysis, decompilation and more, given a large set of plugins.
```

> Since version 8 Angr is based on Python 3, and it's available from PyPI. With pip, it's easy to install on \*nix operating systems and Mac OS:

Reviewer: ```suggestion
Since version 8, Angr is based on Python 3, and can be installed with pip on \*nix operating systems, macOS and Windows:
```

```shell

> $ mkvirtualenv --python=$(which python3) angr && python -m pip install angr

Reviewer: I'd put the virtualenv stuff in a note and just leave "pip install angr" here.

```suggestion
$ pip install angr
```

```

> Creating a dedicated virtual environment with Virtualenv is recommended because some of its dependencies contain forked versions Z3 and PyVEX, which overwrite the original versions. You can skip this step if you don't use these libraries for anything else. Alternatively, there is always a docker container which you can use.

Reviewer: As said in the prev. comment:

```suggestion
> Some of angr's dependencies contain forked versions of the Python modules Z3 and PyVEX, which would overwrite the original versions. If you're using those modules for anything else, you should create a dedicated virtual environment with Virtualenv. Alternatively, you can always use the provided docker container. See the [installation guide](https://docs.angr.io/introductory-errata/install "angr Installation Guide") for more details.
```

> Comprehensive documentation, including an installation guide, tutorials, and usage examples is available on [Gitbooks page of angr](https://docs.angr.io/ "angr"). A complete [API reference](https://angr.io/api-doc/ "angr API") is also available.

Reviewer: ```suggestion
Comprehensive documentation, including an installation guide, tutorials, and usage examples are available on [Gitbooks page of angr](https://docs.angr.io/ "angr"). A complete [API reference](https://angr.io/api-doc/ "angr API") is also available.
```

> You can use angr with a pythorn REPL - such as iPython - or script your approaches. Although angr has a bit of a steep learning curve, we do recommend using it when you want to brute force your way to a given state of an executable. See [Binary Analysis Frameworks](###Binary Analysis Frameworks) as a great example of how this can work.

Reviewer: Fixes typo, section name (actually is Binary Analysis now, but angr is in Symbolic Execution), also referencing with several #'s doesn't work. only 1 is required

```suggestion
You can use angr from a Python REPL - such as iPython - or script your approaches. Although angr has a bit of a steep learning curve, we do recommend using it when you want to brute force your way to a given state of an executable. See section [Symbolic Execution](#Symbolic-Execution) as a great example on how this can work.
```
